lib/datadog/statsd/sender.rb

```diff
@@ -2,70 +2,111 @@
 
 module Datadog
   class Statsd
+    # Sender is using a companion thread to flush and pack messages
+    # in a `MessageBuffer`.
+    # The communication with this thread is done using a `Queue`.
+    # If the thread is dead, it is starting a new one to avoid having a blocked
+    # Sender with no companion thread to communicate with (most of the time, having
+    # a dead companion thread means that a fork just happened and that we are
+    # running in the child process).
     class Sender
       CLOSEABLE_QUEUES = Queue.instance_methods.include?(:close)
 
-      def initialize(message_buffer)
+      def initialize(message_buffer, logger: nil)
         @message_buffer = message_buffer
+        @logger = logger
+        @mx = Mutex.new
       end
 
       def flush(sync: false)
-        # don't try to flush if there is no message_queue instantiated
-        return unless message_queue
+        # don't try to flush if there is no message_queue instantiated or
+        # no companion thread running
+        if !message_queue
+          @logger.debug { "Statsd: can't flush: no message queue ready" } if @logger
+          return
+        end
+        if !sender_thread.alive?
+          @logger.debug { "Statsd: can't flush: no sender_thread alive" } if @logger
+          return
+        end
 
-        message_queue.push(:flush)
+        @mx.synchronize {
+          message_queue.push(:flush)
```

Comment:
Note: If we get here, there's no guarantee that `message_queue` is non-nil, because another thread might've called `stop` between our previous check and now. I suggest doing something like

```ruby
current_message_queue = message_queue
```

at the beginning of the method, and then always use this local copy.

(Arguably, if people are calling stop and flush concurrently they're setting themselves up for some pain BUT this change is simple and ensures that we don't cause a weird NoMethodError from this line)

Reply:
@ivoanjo Can I do this instead?

```ruby
          message_queue.push(:flush) if message_queue
```

Reply:
Yup, not atomic either. In from Ruby 2.3+ you can do `message_queue&.push(:flush)` (and this would only read `message_queue` once), but I see we're still supporting 2.0 (dd-trace-rb is actually at 2.1 minimum), I think you really need to introduce the local variable.